src/core/components/pin-set.js

```diff
@@ -0,0 +1,271 @@
+  }
+  return {
+    header: header,
+    data: rootData.slice(hdrLength + vBytes)
+  }
+}
+
+function hash (seed, key) {
+  const buf = Buffer.alloc(4)
+  buf.writeUInt32LE(seed, 0)
+  const data = Buffer.concat([
+    buf, Buffer.from(toB58String(key))
+  ])
+  return fnv1a(data.toString('binary'))
+}
+
+exports = module.exports = function (dag) {
+  const pinSet = {
+    // should this be part of `object` API?
+    hasChild: (root, childhash, callback) => {
+      let seen = {}
+      if (typeof childhash === 'object') {
+        childhash = toB58String(childhash)
+      }
+
+      function searchChildren (root, cb, _links, _checked) {
+        if (!root.links.length && _links === _checked) {
+          // all nodes have been checked
+          return cb(null, false)
+        }
+
+        some(root.links, (link, cb) => {
+          const bs58Link = toB58String(link.multihash)
+          if (bs58Link in seen) return
+          if (bs58Link === childhash) {
+            return cb(null, true)
+          }
+
+          seen[bs58Link] = true
+
+          dag.get(link.multihash, (err, res) => {
+            if (err) return cb(err)
+
+            _checked++
+            _links += res.value.links.length
+            searchChildren(res.value, cb, _links, _checked)
+          })
+        }, cb)
+      }
+
+      return searchChildren(root, callback, root.links.length, 0)
+    },
+
+    storeSet: (keys, logInternalKey, callback) => {
+      const items = keys.map(key => ({
+        key: key,
+        data: null
+      }))
+
+      pinSet.storeItems(items, logInternalKey, (err, rootNode) => {
+        if (err) { return callback(err) }
+        const opts = { cid: new CID(rootNode.multihash) }
+        dag.put(rootNode, opts, (err, cid) => {
+          if (err) { return callback(err) }
+          logInternalKey(rootNode.multihash)
+          callback(null, rootNode)
+        })
+      })
+    },
+
+    storeItems: (items, logInternalKey, callback, _depth, _subcalls, _done) => {
+      callback = once(callback)
+      const pbHeader = pb.Set.encode({
+        version: 1,
+        fanout: defaultFanout,
+        seed: _depth
+      })
+      let rootData = Buffer.concat([
+        Buffer.from(varint.encode(pbHeader.length)), pbHeader
+      ])
+      let rootLinks = []
+      for (let i = 0; i < defaultFanout; i++) {
+        rootLinks.push(new DAGLink('', 1, emptyKey))
+      }
+      logInternalKey(emptyKey)
+
+      if (items.length <= maxItems) {
+        // the items will fit in a single root node
+        const itemLinks = []
+        const itemData = []
+        const indices = []
+
+        for (let i = 0; i < items.length; i++) {
+          itemLinks.push(new DAGLink('', 1, items[i].key))
+          itemData.push(items[i].data || Buffer.alloc(0))
+          indices.push(i)
+        }
+        indices.sort((a, b) => {
+          const x = Buffer.compare(itemLinks[a].multihash, itemLinks[b].multihash)
+          if (x) { return x }
```

Comment:
Do we need to sort the links? Imagine directories that have over a million links, can we reasonably sort them or should we ignore sorting like go-ipfs does?

Reply:
Thanks guys, I will check go's handling and run tests then undo the sorting if we can get away with it :). Pin's performance isn't stellar so removing this if we can would be good.

Reply:
So go [does sort](https://github.com/JonKrone/go-ipfs/blob/master/pin/set.go#L97) the links for the DAGNode. I don't know why it's important though, do you know what impact sorting the links of a DAGNode could have? I can ask around irc as well.

It also turns out that our sorting never ran because the indices array was always already sorted (d'oh). Even without sorting, the pin interop tests show go and js produce the same pins.

Reply:
@VictorBjelkholm The links are sorted so that the order of the pins doesn't affect the hash of the `DAGNode` we create and then `put` in the datastore. So storing pins `{A, B, C}` produces the same root node as storing pins `{B, A, C}` Added a comment to log this :).